frontend/src/util/compile.ts

```diff
@@ -1,37 +1,151 @@
+   * @param machine The current wasm machine
+   * @param runtimeError The current time error, if one has been encountered
+   * @return A state object fit for React consumption
+   */
+  private static getMachineState(
+    machine: Machine,
+    runtimeError: SourceElement | undefined
+  ): MachineState {
+    return {
+      programCounter: machine.programCounter,
+      input: Array.from(machine.input),
+      output: Array.from(machine.output),
+      registers: machine.registers,
+      stacks: machine.stacks,
+      cycleCount: machine.cycleCount,
+      isComplete: machine.isComplete,
+      isSuccessful: machine.isSuccessful,
+      runtimeError,
+    };
+  }
+
+  /**
+   * Helper to refresh the current state based on the current wasm values.
+   */
+  private updateState(): void {
+    this._state = MachineWrapper.getMachineState(
+      this.machine,
+      this.runtimeError
+    );
+  }
+
+  /**
+   * Execute the next instruction. This will update the machine state after
+   * the step is run to reflect the new state.
+   */
+  executeNext(): void {
+    try {
+      this.machine.executeNext();
+    } catch (e) {
+      // Make sure this is the error type we expect
+      if (isSourceElement(e)) {
+        this.runtimeError = e;
+      } else {
+        // Unrecognized error, throw it back up
+        throw e;
+      }
+    }
+    this.updateState();
+  }
 
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-export type Instruction = any;
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-export type Error = any;
+  get state(): MachineState {
+    return this._state;
+  }
+}
 
-type CompileRes = 'ERROR: Invalid Hardware Spec' | CompiledRes;
-export default class Compiler {
+export type CompileResult =
+  | { type: 'compiled'; instructions: SourceElement[]; machine: MachineWrapper }
+  | { type: 'error'; errors: SourceElement[] };
+
+export class CompilerWrapper {
   static gdlk: {
     compile: typeof compile;
-    HardwareSpec: typeof HardwareSpec;
   };
 
   static async init(): Promise<void> {
-    Compiler.gdlk = await import('gdlk_wasm');
+    CompilerWrapper.gdlk = await import('gdlk_wasm');
   }
 
-  static compile(input: string, hwSpec: HardwareSpec): CompileRes {
-    return Compiler.gdlk.compile(input, hwSpec);
-  }
+  static compile(
+    hardwareSpec: HardwareSpec,
+    programSpec: ProgramSpec,
+    source: string
+  ): CompileResult {
+    try {
+      const result = CompilerWrapper.gdlk.compile(
+        hardwareSpec,
+        programSpec,
+        source
+      );
 
-  static makeHardwareSpec = (
-    num_registers: number,
-    num_stacks: number,
-    max_stack_length: number
-  ): HardwareSpec => {
-    return new Compiler.gdlk.HardwareSpec(
-      num_registers,
-      num_stacks,
-      max_stack_length
-    );
-  };
+      return {
+        type: 'compiled',
+        // Do NOT try to change this to a spread!
```

Comment:
?

Reply:
ah this makes less sense now, i'll remove the comment